mapeditor/campaigneditor/campaigneditor.cpp

```diff
@@ -0,0 +1,260 @@
+
+	auto background = BitmapHandler::loadBitmap(campaignState->getRegions().getBackgroundName().getName());
+	if(!ui->actionShowFullBackground->isChecked())
+		background = background.copy(0, 0, 456, 600); 
+	campaignScene->addItem(new QGraphicsPixmapItem(QPixmap::fromImage(background)));
+	for (auto & s : campaignState->scenarios)
+	{
+		auto scenario = s.first;
+		auto color = campaignState->scenarios.at(scenario).regionColor;
+		auto image = BitmapHandler::loadBitmap(campaignState->getRegions().getAvailableName(scenario, color).getName());
+		if(selectedScenario == scenario)
+			image = BitmapHandler::loadBitmap(campaignState->getRegions().getSelectedName(scenario, color).getName());
+		else if(campaignState->scenarios.at(scenario).mapName == "")
+			image = BitmapHandler::loadBitmap(campaignState->getRegions().getConqueredName(scenario, color).getName());
+		auto pixmap = new ClickablePixmapItem(QPixmap::fromImage(image), [this, scenario]()
+		{
+			bool redrawRequired = selectedScenario != scenario;
+			selectedScenario = scenario;
+
+			if(redrawRequired)
+				redraw();
+		}, [this, scenario]()
+		{
+			if(ScenarioProperties::showScenarioProperties(campaignState, scenario))
+				changed();
+			redraw();
+		}, [this, scenario](QGraphicsSceneContextMenuEvent * event)
+		{
+			QMenu contextMenu(this);
+			QAction *actionScenarioProperties = contextMenu.addAction(tr("Scenario editor"));
+			actionScenarioProperties->setIcon(ui->actionScenarioProperties->icon());
+			connect(actionScenarioProperties, &QAction::triggered, this, [this, scenario]() {
+				if(ScenarioProperties::showScenarioProperties(campaignState, scenario))
+					changed();
+				redraw();
+			});
+			contextMenu.exec(event->screenPos());
+		});
+		auto pos = campaignState->getRegions().getPosition(scenario);
+		pixmap->setPos(pos.x, pos.y);
+		pixmap->setToolTip(QString::fromStdString(campaignState->scenarios.at(scenario).mapName));
+		campaignScene->addItem(pixmap);
+	}
+
+	campaignScene->setSceneRect(background.rect());
+	ui->campaignView->show();
+}
+
+bool CampaignEditor::getAnswerAboutUnsavedChanges()
+{
+	if(unsaved)
+	{
+		auto sure = QMessageBox::question(this, tr("Confirmation"), tr("Unsaved changes will be lost, are you sure?"));
+		if(sure == QMessageBox::No)
+		{
+			return false;
+		}
+	}
+	return true;
+}
+
+void CampaignEditor::setTitle()
+{
+	QFileInfo fileInfo(filename);
+	QString title = QString("%1%2 - %3 (%4)").arg(fileInfo.fileName(), unsaved ? "*" : "", tr("VCMI Campaign Editor"), GameConstants::VCMI_VERSION.c_str());
+	setWindowTitle(title);
+}
+
+void CampaignEditor::changed()
+{
+	unsaved = true;
+	setTitle();
+}
+
+bool CampaignEditor::saveCampaign()
+{
+	if(campaignState->mapPieces.size() != campaignState->campaignRegions.regions.size())
+	{
+		auto reply = QMessageBox::question(nullptr, tr("Maps missing"), tr("Not all regions have a map. Do you want to continue?"), QMessageBox::Yes|QMessageBox::No);
+		if (reply != QMessageBox::Yes)
+			return false;
```

Comment:
I think this is legal and common in user-made campaigns. So no need to ask? Or I am wrong?

Reply:
Absolutely normal to have non used regions.